File: src/Random.cpp
/*
  Copyright (C) 2009  Jon Macey

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITH ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "Random.h"
#include <cstdlib> // for time
#include <ctime>

//----------------------------------------------------------------------------------------------------------------------
/// @file Random.cpp
/// @brief implementation files for Random class
//----------------------------------------------------------------------------------------------------------------------
namespace ngl
{
constexpr auto RandomFloat = "RandomFloat";
constexpr auto RandomPositiveFloat = "RandomPositiveFloat";
std::mt19937 Random::m_generator;
// because of C++ 11 we can now use lambda expressions to init static members so we don't need a ctor
std::unordered_map< std::string, std::uniform_real_distribution< Real > > Random::m_floatGenerators = []
{
  std::unordered_map< std::string, std::uniform_real_distribution< Real > > gen;
  // first create a simple uniform real distrib
  std::uniform_real_distribution< Real > MinusPlusOneFloatDistrib(-1.0f, 1.0f);
  gen[RandomFloat] = MinusPlusOneFloatDistrib;
  // same for below but using 0-1 for distrib
  std::uniform_real_distribution< Real > ZeroOneFloatDistrib(0.0f, 1.0f);
  gen[RandomPositiveFloat] = ZeroOneFloatDistrib;
  return gen;
}();

std::unordered_map< std::string, std::uniform_int_distribution< int > > Random::m_intGenerators;

//----------------------------------------------------------------------------------------------------------------------
void Random::setSeed()
{
  m_generator.seed(static_cast< unsigned int >(std::time(nullptr)));
}

//----------------------------------------------------------------------------------------------------------------------
void Random::setSeed(unsigned int _value)
{
  m_generator.seed(_value);
}

//----------------------------------------------------------------------------------------------------------------------
Real Random::getFloatFromGeneratorName(std::string_view _name)
{
  // grab a function pointer based on the _name from the map
  // auto func=m_floatGenerators[_name];
  auto func = m_floatGenerators.find(_name.data());
  // see if we got anything we can use
  if(func != m_floatGenerators.end())
  {
    // if it exists execute the function and return the value
    return func->second(m_generator);
  }
  else
  {
    // otherwise we return the safest possible value 0
    return 0.0f;
  }
}

int Random::getIntFromGeneratorName(std::string_view _name)
{
  // grab a function pointer based on the _name from the map
  // auto func=m_floatGenerators[_name];
  auto func = m_intGenerators.find(_name.data());
  // see if we got anything we can use
  if(func != m_intGenerators.end())
  {
    // if it exists execute the function and return the value
    return func->second(m_generator);
  }
  else
  {
    // otherwise we return the safest possible value 0
    return 0;
  }
}

void Random::addIntGenerator(std::string_view _name, std::uniform_int_distribution< int > &_dist)
{
  m_intGenerators[_name.data()] = _dist;
}

void Random::addFloatGenerator(std::string_view _name, std::uniform_real_distribution< float > &_dist)
{
  m_floatGenerators[_name.data()] = _dist;
}
//----------------------------------------------------------------------------------------------------------------------
Vec4 Random::getRandomVec4()
{
  auto gen = m_floatGenerators[RandomFloat];
  return Vec4(gen(m_generator), gen(m_generator), gen(m_generator), 0.0f);
}

Vec4 Random::getRandomColour4()
{
  auto gen = m_floatGenerators[RandomPositiveFloat];
  return Vec4(gen(m_generator), gen(m_generator), gen(m_generator), 1.0f);
}
Vec3 Random::getRandomColour3()
{
  auto gen = m_floatGenerators[RandomPositiveFloat];
  return Vec3(gen(m_generator), gen(m_generator), gen(m_generator));
}

//----------------------------------------------------------------------------------------------------------------------
Vec4 Random::getRandomNormalizedVec4()
{
  auto gen = m_floatGenerators[RandomFloat];
  Vec4 v(gen(m_generator), gen(m_generator), gen(m_generator), 0.0f);
  v.normalize();
  return v;
}

//----------------------------------------------------------------------------------------------------------------------
Vec3 Random::getRandomVec3()
{
  auto gen = m_floatGenerators[RandomFloat];
  return Vec3(gen(m_generator), gen(m_generator), gen(m_generator));
}

//----------------------------------------------------------------------------------------------------------------------
Vec3 Random::getRandomNormalizedVec3()
{
  auto gen = m_floatGenerators[RandomFloat];
  Vec3 v(gen(m_generator), gen(m_generator), gen(m_generator));
  v.normalize();
  return v;
}

//----------------------------------------------------------------------------------------------------------------------
Vec2 Random::getRandomVec2()
{
  auto gen = m_floatGenerators[RandomFloat];
  return Vec2(gen(m_generator), gen(m_generator));
}

//----------------------------------------------------------------------------------------------------------------------
Vec2 Random::getRandomNormalizedVec2()
{
  auto gen = m_floatGenerators[RandomFloat];
  Vec2 v(gen(m_generator), gen(m_generator));
  v.normalize();
  return v;
}

Vec3 Random::getRandomPoint(Real _xRange, Real _yRange, Real _zRange)
{
  auto gen = m_floatGenerators[RandomFloat];
  return Vec3(gen(m_generator) * _xRange, gen(m_generator) * _yRange, gen(m_generator) * _zRange);
}

//----------------------------------------------------------------------------------------------------------------------
Real Random::randomNumber(Real _mult)
{
  auto gen = m_floatGenerators[RandomFloat];
  return gen(m_generator) * _mult;
}

//----------------------------------------------------------------------------------------------------------------------
Real Random::randomPositiveNumber(Real _mult)
{
  auto gen = m_floatGenerators[RandomPositiveFloat];
  return gen(m_generator) * _mult;
}
} // namespace ngl

//----------------------------------------------------------------------------------------------------------------------

File: py-ngl/src/ngl/random.py
"""
A module to replicate the functionality of the NGL C++ Random class.
"""

import random
import time

from ngl.vec2 import Vec2
from ngl.vec3 import Vec3
from ngl.vec4 import Vec4


class Random:
    _float_generators = {
        "RandomFloat": lambda: random.uniform(-1.0, 1.0),
        "RandomPositiveFloat": lambda: random.uniform(0.0, 1.0),
    }

    _int_generators = {}

    @staticmethod
    def set_seed():
        """set the seed using std::time(NULL)"""
        random.seed(int(time.time()))

    @staticmethod
    def set_seed_value(value: int):
        """set the seed using a param value
        Args:
            value (int): the seed value
        """
        random.seed(value)

    @staticmethod
    def get_float_from_generator_name(name: str) -> float:
        """gets a pre-generated float value for a genetator
        Args:
            name (str): the name of the generator to use for the number
        Returns:
            a random number created by the generator or 0 if the generator is not found
        """
        if name in Random._float_generators:
            return Random._float_generators[name]()
        return 0.0

    @staticmethod
    def get_int_from_generator_name(name: str) -> int:
        """gets a pre-generated int value for a genetator
        Args:
            name (str): the name of the generator to use for the number
        Returns:
            a random number created by the generator or 0 if the generator is not found
        """
        if name in Random._int_generators:
            return Random._int_generators[name]()
        return 0

    @staticmethod
    def add_int_generator(name: str, generator):
        """add a generator to the int generators
        Args:
            name (str): the name of the generator to use for the number
            generator : the generator to add should be a callable function
        """
        Random._int_generators[name] = generator

    @staticmethod
    def add_float_generator(name: str, generator):
        """add a generator to the float generators
        Args:
            name (str): the name of the generator to use for the number
            generator : the generator to add should be a callable function
        """
        Random._float_generators[name] = generator

    @staticmethod
    def get_random_vec4() -> Vec4:
        """get a random vector with componets ranged from +/- 1"""
        gen = Random._float_generators["RandomFloat"]
        return Vec4(gen(), gen(), gen(), 0.0)

    @staticmethod
    def get_random_colour4() -> Vec4:
        """get a random colour with components ranged from 0-1"""
        gen = Random._float_generators["RandomPositiveFloat"]
        return Vec4(gen(), gen(), gen(), 1.0)

    @staticmethod
    def get_random_colour3() -> Vec3:
        """get a random colour with components ranged from 0-1"""
        gen = Random._float_generators["RandomPositiveFloat"]
        return Vec3(gen(), gen(), gen())

    @staticmethod
    def get_random_normalized_vec4() -> Vec4:
        """get a random vector with componets ranged from +/- 1 and Normalized"""
        gen = Random._float_generators["RandomFloat"]
        v = Vec4(gen(), gen(), gen(), 0.0)
        v.normalize()
        return v

    @staticmethod
    def get_random_vec3() -> Vec3:
        """get a random vector with componets ranged from +/- 1"""
        gen = Random._float_generators["RandomFloat"]
        return Vec3(gen(), gen(), gen())

    @staticmethod
    def get_random_normalized_vec3() -> Vec3:
        """get a random vector with componets ranged from +/- 1 and Normalized"""
        gen = Random._float_generators["RandomFloat"]
        v = Vec3(gen(), gen(), gen())
        v.normalize()
        return v

    @staticmethod
    def get_random_vec2() -> Vec2:
        """get a random vector with componets ranged from +/- 1"""
        gen = Random._float_generators["RandomFloat"]
        return Vec2(gen(), gen())

    @staticmethod
    def get_random_normalized_vec2() -> Vec2:
        """get a random vector with componets ranged from +/- 1 and Normalized"""
        gen = Random._float_generators["RandomFloat"]
        v = Vec2(gen(), gen())
        v.normalize()
        return v

    @staticmethod
    def get_random_point(x_range: float = 1.0, y_range: float = 1.0, z_range: float = 1.0) -> Vec3:
        """get a random point in 3D space defaults to +/- 1 else user defined range
        Args:
            x_range (float): the +/-x range
            y_range (float): the +/-y range
            z_range (float): the +/-z range
        Returns:
            a random point
        """
        gen = Random._float_generators["RandomFloat"]
        return Vec3(gen() * x_range, gen() * y_range, gen() * z_range)

    @staticmethod
    def random_number(mult: float = 1.0) -> float:
        """a replacement for the old RandomNumber func
        this is basically a convinience function
        Args:
            mult (float): an optional multiplyer for the output
        Returns:
            (uniform_random(-1-0-+1) * mult)
        """
        gen = Random._float_generators["RandomFloat"]
        return gen() * mult

    @staticmethod
    def random_positive_number(mult: float = 1.0) -> float:
        """a replacement for the old ReandomPosNum
        this is basically a convinience function
        Args:
            mult (float): an optional multiplyer for the output
        Returns:
            (uniform_random(0-1) * mult)
        """
        gen = Random._float_generators["RandomPositiveFloat"]
        return gen() * mult
